src/System/Terminal/Widgets/SearchSelect.hs

```diff
@@ -0,0 +1,126 @@
+    -- ^ The options presented to a user. Searched through via the searchField.
+    -- This widget should /not/ be used to query all available options;
+    -- it only shows the filtered output.
+    , globalOptions :: Map Text Bool
+    -- ^ Used internally to keep track of all options and their statuses, even
+    -- when not shown to the user.
+    , showOnNoInput :: Bool
+    -- ^ Whether to show the options when the user has yet to type any input.
+    , showCount :: Int
+    -- ^ The maximum number of items to show at any given time.
+    }
+    deriving stock (Generic)
+
+filterSS :: Text -> SearchSelect -> SearchSelect
+filterSS t ss@(SearchSelect{selectWidget, globalOptions, showCount}) =
+    ss{selectWidget = newSelect}
+  where
+    filteredEntries =
+        take showCount
+            . Map.toList
+            $ Map.filterWithKey (const . T.isInfixOf t) globalOptions
+    filterLen = length filteredEntries
+    newCursorRow
+        | selectWidget.cursorRow > filterLen =
+            -- 'max' prevents hookups, since 'Select' is 1-indexed
+            max filterLen 1
+        | otherwise = selectWidget.cursorRow
+    newSelect = selectWidget{options = filteredEntries, cursorRow = newCursorRow}
+
+updateGlobal :: Select -> Map Text Bool -> Map Text Bool
+updateGlobal sel
+    | sel.multiselect = Map.union updates
+    | otherwise = Map.union updates . Map.map (const False)
+  where
+    updates = Map.fromList $ filter snd sel.options
+
+searchVal :: TextInput -> Text
+searchVal = RopeZipper.toText . (.value)
+
+highlightSelected :: Select -> Select
+highlightSelected s@(Select{options, cursorRow}) = s{options = newOptions}
+  where
+    i = max (cursorRow - 1) 0
+    (t, b) = options !! i
+    stylizedSelOpt = ("[" <> t <> "]", b)
+    newOptions
+        | null options = []
+        | otherwise = take i options ++ [stylizedSelOpt] ++ drop (i + 1) options
+
+instance Widget SearchSelect where
+    cursor :: Lens' SearchSelect Position
+    cursor = lens getter setter
+      where
+        getter :: SearchSelect -> Position
+        getter s = s ^. #searchWidget . cursor
+        setter :: SearchSelect -> Position -> SearchSelect
+        setter s p = s & #searchWidget . cursor .~ p
+    handleEvent (KeyEvent EnterKey []) ss@(SearchSelect{selectWidget, globalOptions})
+        | null selectWidget.options = ss
+        | otherwise =
+            let newSelect = handleEvent (KeyEvent SpaceKey []) selectWidget
+                newGlobal = updateGlobal newSelect globalOptions
+             in ss{selectWidget = newSelect, globalOptions = newGlobal}
+    handleEvent ev@(KeyEvent (ArrowKey _) []) ss@(SearchSelect{selectWidget}) =
+        let newSelect = handleEvent ev selectWidget
+         in ss{selectWidget = newSelect}
+    handleEvent ev ss@(SearchSelect{searchWidget}) =
+        let newSearch = handleEvent ev searchWidget
+            search = searchVal newSearch
+         in (filterSS search ss){searchWidget = newSearch}
```

Comment:
Replace pattern match with `#selectWidget %~`.

Reply:
i fixed it for the first instance here, but the second instance is sorta awkward; we want to handle the event *first*, then filter, then apply. this order of operations is awkward with lenses, since we want to apply the changes in a specified order rather than all at once. That or my lens-foo isn't good enough to nicely solve this particular instance